src/clj_kondo/impl/analyzer.clj

```diff
@@ -486,10 +490,11 @@
               ctx)
         private? (or (= "defn-" call)
                      (:private var-meta))
-        docstring (or (some-> meta-node2-meta :doc str)
-                      (some-> meta-node-meta :doc str)
-                      docstring
-                      (some-> var-leading-meta :doc str))
+        [doc-node docstring] (if docstring
+                               [doc-node docstring]
+                               ;; TODO: too late to get raw var-leading-meta node
```

Comment:
The doc-node locals are shadowing each other and that probably relates to this TODO.

Reply:
The shadowing is intentional here and in the other spots - in general docstrings can be specified as either 

 - string in the correct position of the ‘def*’ expression. 
 - :doc entry in attr map literal
 - :doc entry in metadata on the _symbol_ naming the var (using ‘^{:doc …} var-name’ reader syntax. 

since there can be only one docstring for a var, the code has to determine what to do if multiple ways of defining the docstring are used.  

This logic was already present  before this patch but I extended the code because the linter needs the ‘doc-node’  _node_ (not just the ‘docstring’ _content_) in order to report on the position of the docstring in the sourcecode. 

The above is compete for all cases except the “metadata on the symbol” case - the analyzer code handles that through  calls to ‘lift-meta-content2’ which drops the raw node information and I could not quickly find a good way to pass that through and I need to play with it a bit more. It’s a pretty dense bit of code and I don’t really get all the nuances of what it’s doing.

My guess is that the right solution here is actually slightly different from the current implementation - instead of calling ‘lint-docstring!’ Immediately from the analyzer, push docstring positon information into the generated metadata map and call lint-docstring when reg-var! Is called. That way you can also do docstring checks later on - for instance to check that vars mentioned in docstrings exist in the namespace.

Reply:
Right now we already had logic to detect which is _the_ docstring and this logic is now kind of duplicated.

> instead of calling ‘lint-docstring!’ Immediately from the analyzer, push docstring positon information into the generated metadata map and call lint-docstring when reg-var! Is called. That way you can also do docstring checks later on - for instance to check that vars mentioned in docstrings exist in the namespace.

Yes, agreed, that would be a much better approach.